integration-cli/docker_api_containers_unix_test.go

```diff
@@ -0,0 +1,49 @@
+// +build !windows
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"syscall"
+
+	"github.com/docker/docker/integration-cli/checker"
+	"github.com/docker/docker/integration-cli/request"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/go-check/check"
+)
+
+func (s *DockerSuite) TestContainerAPIArchiveWithHardlink(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+
+	containerID := makeTestContainer(c, testContainerOptions{addContent: true})
+
+	tmpDir := getTestDir(c, "test-cp-from-dir-with-hardlink")
+	defer os.RemoveAll(tmpDir)
+
+	srcPath := "/dir6"
+	dstPath := cpPath(tmpDir, "dir6_")
+	dstFile := cpPath(tmpDir, "dir6_/hardLinkFile1")
+	endpoint := fmt.Sprintf("/containers/%s/archive?path=%s", containerID, srcPath)
+	res, body, err := request.SockRequestRaw("GET", endpoint, nil, "application/x-tar", daemonHost())
+	c.Assert(err, checker.IsNil)
+	c.Assert(res.StatusCode, checker.Equals, http.StatusOK)
+
+	dstDir, dstBase := archive.SplitPathDirEntry(dstPath)
+	_, srcBase := archive.SplitPathDirEntry(srcPath)
+	copyArchive := archive.RebaseArchiveEntries(body, srcBase, dstBase)
```

Comment:
@dnephin thanks for your reply, I 'a not sure whether I understand you completely, Let me show what I think.

> docker cp id:/dir6 /tmpxxx/_dir6

the case has the following feathres,
- there is hardlink in  id:/dir6 
- the dir "_dir6" is not exist.
- the baseDirs(dir6 and _dir6) are not the same, old is dir6, new is dir_, dir6 != _dir6

before https://github.com/moby/moby/pull/33956, RebaseArchiveEntries has a bug, it don't change base name, it will lead the following issue,

"link /tmp/test-cp-from-dir-with-hardlink085642536/**dir6**/hardFile1 /tmp/test-cp-from-dir-with-hardlink085642536/**dir6_**/hardLinkFile1: no such file or directory"

the case above can reproduce the issue before https://github.com/moby/moby/pull/33956.

the correct should be,
"link /tmp/test-cp-from-dir-with-hardlink085642536/**dir6_**/hardFile1 /tmp/test-cp-from-dir-with-hardlink085642536/**dir6_**/hardLinkFile1"

and https://github.com/moby/moby/pull/33956 fix it. So RebaseArchiveEntries()  is called here.